### ISSUE TITLE
isolation_lavel='autocommit' causes exception on next transaction.

### ISSUE THREAD

**Issue by evgen231** [sqlalchemy/sqlalchemy#7141]
### Describe the bug

SQLAlchemy raises error on next transaction after isolation_level="AUTOCOMMIT" query. It affects "future engines" (`create_engine(..., future=True)`).

### To Reproduce

```python
import sqlalchemy as sa

engine = sa.create_engine('postgresql://...',
                          echo=True, future=True)
with engine.connect() as connection:
    connection \
        .execution_options(isolation_level='AUTOCOMMIT') \
        .execute(sa.text("select 1;"))

    with connection.begin():
        connection.execute(sa.text('select 2'))
```

### Error

```
# Copy complete stack trace and error message here, including SQL log output if applicable.
2021-10-05 01:34:44,427 INFO sqlalchemy.engine.Engine select pg_catalog.version()
2021-10-05 01:34:44,428 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-10-05 01:34:44,432 INFO sqlalchemy.engine.Engine select current_schema()
2021-10-05 01:34:44,432 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-10-05 01:34:44,436 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2021-10-05 01:34:44,436 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-10-05 01:34:44,440 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-10-05 01:34:44,440 INFO sqlalchemy.engine.Engine select 1;
2021-10-05 01:34:44,441 INFO sqlalchemy.engine.Engine [generated in 0.00026s] {}
2021-10-05 01:34:44,443 INFO sqlalchemy.engine.Engine ROLLBACK using DBAPI connection.rollback(), DBAPI should ignore due to autocommit mode
Traceback (most recent call last):
  File ".../main.py", line 10, in <module>
    with connection.begin():
  File ".../venv/lib/python3.9/site-packages/sqlalchemy/future/engine.py", line 144, in begin
    return super(Connection, self).begin()
  File ".../venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 766, in begin
    raise exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: a transaction is already begun for this connection

Process finished with exit code 1

``` 

### Versions

- OS: ArchLinux
- Python: 3.9.7
- SQLAlchemy: 1.4.25
- Database: PostgreSQL
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): psycopg with future=True and asyncpg

### Additional context

This bug affects alembic `.autocommit_block()` with the same error.
Looks like same issue https://github.com/sqlalchemy/sqlalchemy/issues/6701.

IMHO the root of the problem is somewere at https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/base.py#L1722. There is no isolation_level check.

**Comment by zzzeek** [comment]
does not appear to be related to #6701 as you are not using the "isolation_level" parameter of create_engine.   the isolation_level execution option necessarily applies to the connection overall and this is documented at https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=execution_options#sqlalchemy.engine.Connection.execution_options.params.isolation_level

>  The isolation level option applies the isolation level by emitting statements on the DBAPI connection, and necessarily affects the original Connection object overall, not just the copy that is returned by the call to Connection.execution_options() method. The isolation level will remain at the given setting until the DBAPI connection itself is returned to the connection pool, i.e. the Connection.close() method on the original Connection is called, where an event handler will emit additional statements on the DBAPI connection in order to revert the isolation level change.

**Comment by zzzeek** [comment]
it seems like you mean to post a bug on alembic.   i can see alembic might not handle this correctly.

**Comment by zzzeek** [comment]
ill post

**Comment by zzzeek** [comment]
https://github.com/sqlalchemy/alembic/issues/944

**Comment by zzzeek** [comment]
you should be able to work around by adding "op.get_bind().commit()" to your migration script

**Comment by evgen231** [comment]
Still getting the error. But works fine with future=False.

```python
import sqlalchemy as sa

engine = sa.create_engine('postgresql://...',
                          echo=True, future=True, isolation_level='AUTOCOMMIT')
with engine.connect() as connection:
    connection.execute(sa.text("select 1;"))
    with connection.begin():
        connection.execute(sa.text('select 2'))
```

```
2021-10-05 04:17:42,086 INFO sqlalchemy.engine.Engine select pg_catalog.version()
2021-10-05 04:17:42,086 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-10-05 04:17:42,088 INFO sqlalchemy.engine.Engine select current_schema()
2021-10-05 04:17:42,088 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-10-05 04:17:42,093 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2021-10-05 04:17:42,093 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-10-05 04:17:42,096 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-10-05 04:17:42,096 INFO sqlalchemy.engine.Engine select 1;
2021-10-05 04:17:42,096 INFO sqlalchemy.engine.Engine [generated in 0.00016s] {}
2021-10-05 04:17:42,098 INFO sqlalchemy.engine.Engine ROLLBACK
Traceback (most recent call last):
  File ".../main.py", line 7, in <module>
    with connection.begin():
  File ".../venv/lib/python3.9/site-packages/sqlalchemy/future/engine.py", line 144, in begin
    return super(Connection, self).begin()
  File ".../venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 766, in begin
    raise exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: a transaction is already begun for this connection

Process finished with exit code 1
```

**Comment by zzzeek** [comment]
1. remove "isolation_level" from create_engine, use only execution options
2. call commit() before next transaction

as follows:

```py
import sqlalchemy as sa

engine = sa.create_engine('postgresql://scott:tiger@localhost/test',
                          echo=True, future=True)
with engine.connect() as connection:
    connection.execution_options(isolation_level="AUTOCOMMIT")
    connection.execute(sa.text("select 1;"))
    connection.commit()
    with connection.begin():
        connection.execute(sa.text('select 2'))
```

**Comment by zzzeek** [comment]
the "isolation_level" doesnt matter, it's still the same behavior.

what the problem is here is that even though we are getting ready to do this whole 2.0 API where nothing is ambiguous and everything will be clear, this looks like a big wave about to hit us of "this will be hard to explain".

1. the connection.begin() method **does not do anything to the database**. it only indicates that the logical start of a "transaction" is here, which will be concluded by commit() or rollback()
2. this means if you are in "AUTOCOMMIT" isolation level, **you stay in autocommit, no transaction is begun**.
3. an additional thing that is confusing the poster here is that **2.0 connections use autobegin**. You can't call "begin()" after running SQL, because the "transaction" was begun already.


basically we need to make sure the docs make **very clear**, somehow, is that the isolation_level execution option is **not in any way related to the begin/commit behavior of the SQLAlchemy connection**, becuase this will keep coming up.

**Comment by CaselIT** [comment]
Worth reopening this to track the documentation update?

**Comment by zzzeek** [comment]
not in this case, I'm doing it now.   i think #6342 is the main problem we have to figure out still and that's a 2.0 thing.

**Comment by zzzeek** [comment]
new section is fairly epic so we will have it close this

**Comment by zzzeek** [comment]
we can do more edits if we need as further commits, this commit is the basic idea
